### PR TITLE
[BEAM-6231] Make Dataflow runner harness work with FixedWindow

### PR DESCRIPTION
**Pull request by boyuanzz**
There is a window error "IntervalWindow cannot be cast to GlobalWindow" introduced by https://github.com/apache/beam/pull/7015. This PR fixes this error and removes all DataflowPortabilityExecutableStageUnsupported tags. 
This test report https://scans.gradle.com/s/645npxy7rmibi/tests is running with
```
includeCategories 'org.apache.beam.testing.DataflowPortabilityExecutableStageUnsupported'
excludeCategories 'org.apache.beam.testing.UsesSideInputs'
excludeCategories 'org.apache.beam.testing.UsesStatefulPardo'
```
R: @robertwb 
cc: @HuangLED

### CODE REVIEW

**Comment by boyuanzz** [comment]
Run Python PreCommit

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java [hunk 1]
```diff
@@ -163,7 +167,7 @@ public Node apply(MutableNetwork<Node, Edge> input) {
       componentsBuilder.putEnvironments(envId, Environments.JAVA_SDK_HARNESS_ENVIRONMENT);
     }
 
-    // Use default WindowingStrategy as the fake one.
+    // Use default WindowingStrategy as the fake one. But it may be overrided later if it's a FixedWindow.
     // TODO: should get real WindowingStategy from pipeline proto.
```
[the remark this answers is not in the document]
Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: I agree and getting pipeline proto from Dataflow service should be an ultimate solution. But it's kind of hard for me to change the proto from Dataflow service, and the proto changes will also take some time. This generated WindowingStrategy is a short term workaround for now. We are discussing the possibility of changing proto from Dataflow service.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java [hunk 2]
```diff
@@ -191,6 +195,37 @@ public Node apply(MutableNetwork<Node, Edge> input) {
...
       InstructionOutput instructionOutput = node.getInstructionOutput();
 
+      // If this is the input PCollection or the output PCollection for an ExecutableStage, it's
+      // necessary to check whether the window coder is not a GlobalWindow coder.
+      if (isExecutableStageInputPCollection(input, node)
+          || isExecutableStageOutputPCollection(input, node)) {
+        Coder<?> javaCoder =
+            CloudObjects.coderFromCloudObject(CloudObject.fromSpec(instructionOutput.getCodec()));
+        // For now, Dataflow runner harness only deal with FixedWindow.
+        if (FullWindowedValueCoder.class.isInstance(javaCoder)) {
```
[the remark this answers is not in the document]
Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: For ```instanceof``` vs ````class.isInstance```, ```instanceof``` checks the type during compiling time but ```class.isInstance``` checks the class of object dynamically (https://www.geeksforgeeks.org/instanceof-operator-vs-isinstance-method-in-java/).

[the remark this answers is not in the document]
Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: For the error, it means there is something wrong when converting a FixWindows into proto. If it fails to convert, then the communication should be broken owing to incomplete proto.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java [hunk 3]
```diff
@@ -191,6 +195,37 @@ public Node apply(MutableNetwork<Node, Edge> input) {
         Iterables.filter(input.nodes(), InstructionOutputNode.class)) {
       InstructionOutput instructionOutput = node.getInstructionOutput();
 
+      // If this is the input PCollection or the output PCollection for an ExecutableStage, it's
```
[the remark this answers is not in the document]
Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: From my understanding, the input and output PCollection go through fnapi data channel between runner harness and sdk harness. Taking sending value from runner harness as an example, it needs a coder(```element coder``` + ```window coder```) to encode value then send it to sdk harness. The ```element coder``` is got from component by ```coderId```, and ```window coder``` is got from WindowingStrategy by ```windowingStrategyId```.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: What I meant was is there any case where javaCoder would not be an instance of FullWindowedValueCoder?

As for isinstance, FullWindowedValueCoder.class is not dynamic, so there's no need to use the less-common form here.  (Same below.)

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: Sorry, let me rephrase this. Is there a reason we should avoid this for non-input-output PCollections?

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java [hunk 4]
```diff
@@ -191,6 +195,37 @@ public Node apply(MutableNetwork<Node, Edge> input) {
...
+              // Since the coder is the only needed from a WindowingStrategy, the size field of one
+              // FixedWindows is meaningless here.
+              RunnerApi.MessageWithComponents windowingStrategyProto =
+                  WindowingStrategyTranslation.toMessageProto(
+                      WindowingStrategy.of(FixedWindows.of(Duration.standardSeconds(1))),
+                      sdkComponents);
+              componentsBuilder.putWindowingStrategies(
+                  fakeWindowingStrategyId, windowingStrategyProto.getWindowingStrategy());
+              componentsBuilder.putAllCoders(windowingStrategyProto.getComponents().getCodersMap());
+              componentsBuilder.putAllEnvironments(
```
Comment (robertwb) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: Do we need this? Wouldn't this break for non-Java?

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java [hunk 5]
```diff
@@ -191,6 +195,37 @@ public Node apply(MutableNetwork<Node, Edge> input) {
...
+        Coder<?> javaCoder =
+            CloudObjects.coderFromCloudObject(CloudObject.fromSpec(instructionOutput.getCodec()));
+        // For now, Dataflow runner harness only deal with FixedWindow.
+        if (FullWindowedValueCoder.class.isInstance(javaCoder)) {
+          FullWindowedValueCoder<?> windowedValueCoder = (FullWindowedValueCoder<?>) javaCoder;
+          Coder<?> windowCoder = windowedValueCoder.getWindowCoder();
+          if (IntervalWindowCoder.class.isInstance(windowCoder)) {
+            fakeWindowingStrategyId = "generatedFixedWindowingStrategy" + idGenerator.getId();
+            try {
+              // Since the coder is the only needed from a WindowingStrategy, the size field of one
```
Comment (robertwb) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: Is this WindowingStrategy instantiated in the SDK? If not, I'd much rather provide an actual fake strategy that has nothing but the coder set (or perhaps a nonce for the URN) to make sure it doesn't actually get used.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: So the v1beta3 protos expose neither the windowing strategy nor a reference that can be used to tie back to the original PCollection? Unfortunate...

In this case, let's add a (possibly memoized) method `getWindowingStrategyForCoder(Coder<?> windowCoder, componentsBuilder)` rather than creating and injecting a default and then later possibly overwriting it. It can have a case for Global and Interval windows, and throw an error otherwise.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: From what I know so far, if there is a window, then FullWindowedValueCoder is the only coder. I feel like the coder I got from v1b3 should be a FullWindowedValueCoder.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: I'm not sure whether I understand the coder usage correctly, but feel like, only the PCollection which is across the sdk-runner boundary cares the window coder. Thus, it's ok just assign the GlobalWindow to these non-input-output PCollections.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: In that case, let's error if not. (Or otherwise discover the exception and properly handle it.)

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: I'd rather not selectively lie, unless there's a good reason to do so.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java (hunk 4 above)
Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: I didn't test these against non-java but I guess there is a huge possibility to break non-java. This is another concern, I may not be able to extract window coder and element coder from a non-java coder.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java (hunk 5 above)
Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: Ideally, the WindowingStrategy should be instantiated in the SDK and passed with pipeline proto to runner harness. However, dataflow runner harness doesn't have pipeline proto thus dataflow generates this fake one in order to make it work with fnapi.

Comment (robertwb) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: Yes, I agree. I'm suggesting we create a very fake one, manually setting the coder_id to the right thing (regardless of whether it's understood by Java), rather than a FixedWindows.

Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: Could you please explain more about how we can manually set the coder_id to the right thing?

Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: Got it. Thanks~

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java [hunk 6]
```diff
@@ -191,6 +195,37 @@ public Node apply(MutableNetwork<Node, Edge> input) {
         Iterables.filter(input.nodes(), InstructionOutputNode.class)) {
       InstructionOutput instructionOutput = node.getInstructionOutput();
 
+      // If this is the input PCollection or the output PCollection for an ExecutableStage, it's
+      // necessary to check whether the window coder is not a GlobalWindow coder.
+      if (isExecutableStageInputPCollection(input, node)
+          || isExecutableStageOutputPCollection(input, node)) {
```
Comment (HuangLED) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: for education purpose, what are the pcollection types beyond Input and output?

**Comment by boyuanzz** [comment]
Run Seed Job

**Comment by boyuanzz** [comment]
commit https://github.com/apache/beam/pull/7356/commits/4f3eaf51db8959e3563fafa786c4305352a08896 addressed the comments above:

- The function will register GlobalWindowingStrategy and FixedWindowingStrategy into componentBuilder before creating all PCollections.
- Instead of only caring about input and output PCollections, it will put the right kind of WindowingStrategy based on window coder to all PCollection. This logic only applies to java coder, for python, it still uses GlobalWindow as default.

Test log: https://scans.gradle.com/s/kameqe3xqgg4u/tests

@robertwb Could you please review this latest changes?

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java (hunk 5 above)
Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: If we don't want to lie on the internal PCollection, then it's not possible to build a totally fake WindowingStrategy. For internal PCollections inside one ExecutanleStage, it needs to initialize WindowFn from WindowingStrategy proto.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java [hunk 7]
```diff
@@ -167,19 +167,24 @@ public Node apply(MutableNetwork<Node, Edge> input) {
       componentsBuilder.putEnvironments(envId, Environments.JAVA_SDK_HARNESS_ENVIRONMENT);
     }
 
-    // Use default WindowingStrategy as the fake one. But it may be overrided later if it's a FixedWindow.
+    // By default, use GlobalWindow for all languages.
+		// For java, if there is a IntervalWindowCoder, then use FixedWindow instead.
```
Comment (robertwb) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: Indentation.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java [hunk 8]
```diff
@@ -167,19 +167,24 @@ public Node apply(MutableNetwork<Node, Edge> input) {
...
     // TODO: should get real WindowingStategy from pipeline proto.
-    String fakeWindowingStrategyId = "fakeWindowingStrategy" + idGenerator.getId();
+    String fakeGlobalWindowingStrategyId = "fakeGlobalWindowingStrategy" + idGenerator.getId();
```
Comment (robertwb) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: Is this one actually fake? Perhaps just call it globalWindowingStrategyId.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java [hunk 9]
```diff
@@ -167,19 +167,24 @@ public Node apply(MutableNetwork<Node, Edge> input) {
...
+    String fakeFixedWindowingStrategyId = "fakeFixedWindowingStrategy" + idGenerator.getId();
```
Comment (robertwb) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: You could call this intervalWindowEncodingWindowingStrategyId or similar to clearly identify the properties you need to leverage here.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java [hunk 10]
```diff
@@ -235,6 +210,17 @@ public Node apply(MutableNetwork<Node, Edge> input) {
           sdkComponents.registerCoder(elementCoder);
           RunnerApi.Coder coderProto = CoderTranslation.toProto(elementCoder, sdkComponents);
           componentsBuilder.putCoders(coderId, coderProto);
+          // For now, Dataflow runner harness only deal with FixedWindow.
+          if (FullWindowedValueCoder.class.isInstance(javaCoder)) {
```
Comment (robertwb) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: You didn't address my comment of why you're not doing `javaCoder isinstance FullWindowedValueCoder`. Generally classObject.isInstrance(Object) is only needed if classObject is not known at compile time.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java [hunk 11]
```diff
@@ -195,38 +200,8 @@ public Node apply(MutableNetwork<Node, Edge> input) {
...
-                  windowingStrategyProto.getComponents().getEnvironmentsMap());
-            } catch (IOException exc) {
-              throw new RuntimeException("Could not convert FixedWindow stratey to proto", exc);
-            }
-          }
-        }
-      }
-
       String coderId = "generatedCoder" + idGenerator.getId();
+      String windowingStrategyId = fakeGlobalWindowingStrategyId;
```
Comment (robertwb) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: I would prefer being defensive and leave this unset, then below setting it in each of the conditions (including adding a check for global windows and the "Coder probably wasn't a java coder" exception clause. This will make it clearer when we're right and when we're just sticking something in.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java (hunk 6 above)
Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: As for an ExecutableStage, there are 3 types of PCollection from my knowledge. The ```input PCollection``` is the PCollection flowing from runner to sdk. The ```output PCollection``` is the PCollection flowing from sdk to runner. The other PCollections are the intermediate PCollection between 2 PTransforms.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java (hunk 11 above)
Comment (boyuanzz) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java]: I don't think we should have an exception clause if this is not a java coder. In my latest changes, I set it to use GlobalWindow if it's not a java coder.

**Comment by boyuanzz** [comment]
Please take another look at commit https://github.com/apache/beam/pull/7356/commits/7857ab0886275d6170f7e189048406baaf2f10a4, @robertwb . Thanks so much for your help!

**Comment by asfgit** [comment]
SUCCESS 
 
--none--

**Comment by boyuanzz** [comment]
Run Java PreCommit

**Comment by boyuanzz** [comment]
Run Java PreCommit
